Integrate CacheSystem inside API-Client

fscarponi opened the pull request:
Introduce caching mechanism inside the PackageSearchApiClient

Added several tests to validate caching mechanisms, including cache hits, cache expiration, and non-exhaustiveness of the cache.

Imported cache library and updated project dependencies. This also includes disabling of certain targets in build configuration.

fscarponi committed:
Importing cache lib, Update dependencies, and disable specific targets

{TMP}Added a new ListPackageSearchEndpoint.kt file to print URLs for various package search API endpoints. Included new repository sources and updated dependency versions in the Gradle build configuration. Also, commented out specific JVM and multiplatform (JS, iOS, macOS, watchOS, tvOS) targets in the build configuration.

fscarponi committed:
WIP

fscarponi committed:
Add HTTP request tests for package search API

Created new `testAPI.http` with multiple endpoints for testing package search API requests, including package info retrieval, known repositories, and package search functionalities. This ensures better coverage and easier testing of different API interactions.

fscarponi committed:
update cache system and add /health endpoint to testApi.http

fscarponi committed:
Update and improve HttpClient's cache system and add matching tests

This update addresses an issue in the HttpClient's cache refresh, wherein previously expired results were not being updated. Changes were made primarily in the fetchPackageInfo method to rectify this.

Additionally, this commit includes new tests for the HttpClient cache system to verify its correct operation post-fix.

fscarponi committed:
Refactor http client and mock engine setup in CacheTests

Abstracted the creation of the mock engine and HTTP client setup to separate helper functions. This reduces redundancy in the test code and improves maintainability. Added test cases for caching logic involving known repositories and package ID hashes.

fscarponi committed:
Add cache behavior tests and refactor setupTestEnv parameter

Introduce multiple tests to verify caching mechanisms including cache hits, cache expiration, and cache non-exhaustiveness. Also, refactor the 'setupTestEnv' function to use 'resourceFilename' parameter instead of 'mockResponseResourceName' for clarity and consistency.

fscarponi committed:
Remove unused functions isOnline and isOffline

The functions isOnline and isOffline were not being used anywhere in the code. Cleaning up these unused functions improves code maintainability and readability.

fscarponi committed:
Refactor fetchPackageInfo and reorganize dependencies

Refactor fetchPackageInfo to handle idHash logic and streamline function calls. Additionally, reorganize test dependencies and Kotlin configurations, including updating the JVM toolchain to version 17.

fscarponi committed:
Refactor cache system and update tests

Removed unnecessary `_id` properties and optimized cache entry updates with `updateWhere` method. Moved test files from `commonTest` to `jvmTest` and ensured all cache-related functionalities are properly tested with adjusted imports.

fscarponi committed:
Refactor cache system and update tests

Removed unnecessary `_id` properties and optimized cache entry updates with `updateWhere` method. Moved test files from `commonTest` to `jvmTest` and ensured all cache-related functionalities are properly tested with adjusted imports.

fscarponi committed:
Remove Gradle cache steps from GitHub Actions

The steps for caching Gradle packages and dependencies have been removed from the test workflow. This change simplifies the workflow and may reduce potential issues related to cache handling. The build step remains unchanged to ensure continuous integration with Gradle.

fscarponi committed:
Update workflow to include cache-integration branch

Modified the GitHub Actions workflow to trigger on pushes to the cache-integration branch. Removed unnecessary blank lines in the script to

fscarponi committed:
Remove unused DateTimeFormatter class and update CI config

Deleted the DateTimeFormatter implementation in the Apple main source set as it was no longer needed. Updated the CI configuration to include a Gradle Enterprise key for the build process.

fscarponi committed:
Remove appleMain dependencies and update CI workflow

Removed the appleMain block in build-systems/build.gradle.kts to streamline the codebase. Modified the GitHub Actions workflow to focus on running specific tests for the cache system and

fscarponi committed:
Update workflow branches and clean up test script

fscarponi committed:
Update workflow branches and clean up test script

fscarponi committed:
Refactor cache query methods to use serializers

Updated cache query methods in `PackageSearchApiClient` to include explicit serializers.

fscarponi committed:
Remove redundant ListPackageSearchEndpoint file

The ListPackageSearchEndpoint file was deleted as it contained a redundant main function that printed URLs already defined in the PackageSearchEndpoints object. This cleanup simplifies the codebase and removes unnecessary duplication.

fscarponi committed:
Remove redundant ListPackageSearchEndpoint file

The ListPackageSearchEndpoint file was deleted as it contained a redundant main function that printed URLs already defined in the PackageSearchEndpoints object. This cleanup simplifies the codebase and removes unnecessary duplication.

fscarponi committed:
wip

fscarponi committed:
Refactor cache handling to improve efficiency

Replaced manual cache updates with generic insert operations and removed unused imports. The changes make the code easier to maintain and extend by leveraging a more consistent cache handling approach. This should also help in reducing the potential for bugs related to cache synchronization.

fscarponi committed:
Refactor cache handling to improve efficiency

- Replaced manual cache updates with generic insert operations
- Removed unused imports
- Improved cache synchronization and maintainability

fscarponi committed:
clarify yml test info

fscarponi committed:
fix

lamba92 commented on build.gradle.kts:
```diff
@@ -10,18 +10,21 @@ kotlin {
                 implementation(packageSearchApiModelsVersions.junit.jupiter.api)
                 implementation(packageSearchApiModelsVersions.junit.jupiter.params)
                 implementation(packageSearchApiModelsVersions.assertk)
+                implementation(kotlinxDocumentStore.mvstore)
```
add also `kotlinxDocumentStore.mvstore` in `jvmMain`, instead of `jvmTest`.

lamba92 commented on http/client/build.gradle.kts:
```diff
@@ -23,10 +25,26 @@ kotlin {
                 api(packageSearchApiModelsVersions.kotlinx.serialization.json)
             }
         }
+        commonTest{
```
indentation here!

lamba92 commented on http/client/build.gradle.kts:
```diff
@@ -23,10 +25,26 @@ kotlin {
             }
         }
+        commonTest{
+            dependencies{
```
indentation here!

lamba92 commented on http/client/src/commonMain/kotlin/cache/CacheEntries.kt:
```diff
@@ -0,0 +1,23 @@
+    val createdAt: Instant = Clock.System.now()
+) {
+
+    internal fun isExpired(expirationTime: Duration = DEFAULT_EXPIRATION_TIME): Boolean =
```
Remove the default here.

lamba92 commented on http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt:
```diff
@@ -1,51 +1,95 @@
+    dataStore: DataStore,
     private val httpClient: HttpClient = defaultHttpClient(),
+    public val endpoints: PackageSearchEndpoints,
+    coroutineScope: CoroutineScope = CoroutineScope(httpClient.coroutineContext),
```
The Ktor `HttpClient` is a coroutine scope itself. Just change this line to:
```kotlin
coroutineScope: CoroutineScope = httpClient
```

lamba92 commented on http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt:
```diff
@@ -146,56 +190,200 @@ public class PackageSearchApiClient(
+            requestBuilder = requestBuilder
+        )
 
     public suspend fun getPackageInfoByIdHashes(
```
This function should be private. Another one without the parameter `useHashes` should be made publicly available:
```kotlin
public suspend fun getPackageInfoByIdHashes(
    idHashes: Set<String>,
    requestBuilder: (HttpRequestBuilder.() -> Unit)? = null,
): Map<String, ApiPackage> = 
    getPackageInfoByIdHashes(
        idHashes = idHashes,
        useHashes = true,
        requestBuilder = requestBuilder
    )
```

lamba92 commented on http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt:
```diff
@@ -146,56 +190,200 @@ public class PackageSearchApiClient(
+
+        val packagesCache = packagesCacheCollection.await()
+
+        val isOffline = !onlineStateFlow.value
```
You can inline this one.
- MacOs shortcut: `cmd` + `option` + `N`
- Windows shortcut: `ctrl` + `alt` + `N`

lamba92 commented on http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt:
```diff
@@ -146,56 +190,200 @@ public class PackageSearchApiClient(
+        scrollCache.find(selector = ScrollRequestCacheEntry::key.name, value = request)
+            .firstOrNull()
+            ?.also {
+                if (!it.isExpired()) return it.value
```
One line it! 
```kotlin
?.also { if (!it.isExpired()) return it.value }
```

lamba92 commented on http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt:
```diff
@@ -211,40 +399,54 @@ public class PackageSearchApiClient(
+
+        val apiProjectsCache = apiProjectCacheCollection.await()
+
+        apiProjectsCache.find(selector = SearchCacheEntry::key.name, value = request).firstOrNull()
```
```kotlin
.find(selector = SearchCacheEntry::key.name, value = request)
   .firstOrNull()
   ?.also { if (!it.isExpired()) return it.value }
```
Please!

lamba92 commented on http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt:
```diff
@@ -211,40 +399,54 @@ public class PackageSearchApiClient(
+                upsert = true,
+                update = ApiProjectCacheEntry(request, it)
+            )
+        }
```
Save also the packages here, just like in `startScroll`.

fscarponi replied:
ApiProject contains only the ids of packages, so we can't

lamba92 replied:
Ah yes, my bad.

lamba92 commented on http/client/src/commonMain/kotlin/org/jetbrains/packagesearch/api/v3/http/PackageSearchApiClient.kt:
```diff
@@ -211,40 +399,54 @@ public class PackageSearchApiClient(
-            requestBuilder = requestBuilder,
-        )
+    ): List<ApiPackage> {
+        //no caches for this endpoint
```
Please put the caches for this as well. Even tho this endpoint does not impact the outbound traffic costs, it still has to be process. Restarting the IDE, re-opening a project, changing the layout of a project and many other actions will still re-trigger this call.

fscarponi committed:
Refactor cache expiration logic and improve code readability

This commit refactors and standardizes cache expiration logic across various functions by making `cacheDuration` an explicit parameter. It also enhances code readability and maintainability by removing redundant variables and ensuring the consistent use of `await` for asynchronous operations. Additionally, it introduces private helper functions and fields to better organize the offline state and cache interaction logic.

fscarponi committed:
adjust apis to not break compatibility
